luaui/Widgets/unit_dgun_stall_assist.lua

```diff
@@ -1,59 +1,88 @@
-
 function widget:GetInfo()
 	return {
-		name      = "DGun Stall Assist",
+		name      = "DGun Stall Assist v2",
 		desc      = "Waits cons/facs when trying to dgun and stalling",
-		author    = "Niobium",
-		date      = "2 April 2010",
+		author    = "zombean",
+		date      = "2 April 2024",
 		license   = "GNU GPL, v2 or later",
 		layer     = 0,
-		enabled   = true  --  loaded by default?
+		enabled   = true
 	}
 end
 
-----------------------------------------------------------------
--- Config
-----------------------------------------------------------------
-local targetEnergy = 600
-local watchForTime = 5
+local millis_start_timer = Spring.GetTimer()
+local function millis()
+  return math.ceil(Spring.DiffTimers(Spring.GetTimer(), millis_start_timer) * 1000)
+end
 
-----------------------------------------------------------------
--- Globals
-----------------------------------------------------------------
-local watchTime = 0
-local waitedUnits = nil -- nil / waitedUnits[1..n] = uID
-local shouldWait = {} -- shouldWait[uDefID] = true / nil
-local isFactory = {} -- isFactory[uDefID] = true / nil
+local target_energy = 1000
+local dgun_cost = 500
+local time_in_ms_befor_unwaiting = 250 --how many fast we can unwait units after waiting them
+local gameStarted = false
+local DGUN_OUT = false
+local UNITS_WAITING = false
+local waitableUnits = {}
+local waitingUnits = {}
+
+local function getWaitAbleUnits()
+  local myWaitableUnits = {}
+  local myUnits = Spring.GetTeamUnits(Spring.GetMyTeamID())
+  for i=1, #myUnits do
+    local unitCmd = Spring.GetUnitCurrentCommand(myUnits[i])
+    local isEnergyConsumingCommand = not (unitCmd == CMD.WAIT or unitCmd == CMD.MOVE or unitCmd == CMD.RECLAIM)
+    if isEnergyConsumingCommand then
+      local unitDefId = Spring.GetUnitDefID(myUnits[i])
+      if waitableUnits[unitDefId] then myWaitableUnits[#myWaitableUnits + 1] = myUnits[i] end
+    end
+  end
+  return myWaitableUnits
+end
 
-local gameStarted
+local function doWeHaveLowEnergy()
+  local currentEnergy, energyStorage = Spring.GetTeamResources(Spring.GetMyTeamID(), "energy")
+  if energyStorage < dgun_cost then return false end
+  if currentEnergy < target_energy then return true end
+  return false
+end
+
+local next_call_when = 0
+local function wait_units()
+  local time_now = millis()
+  if next_call_when > time_now then return end
+  next_call_when = time_now + time_in_ms_befor_unwaiting
+  UNITS_WAITING = true
+  for _, unitID in pairs(getWaitAbleUnits()) do
+    waitingUnits[#waitingUnits +1] = unitID
+  end
+  Spring.GiveOrderToUnitArray(waitingUnits, CMD.WAIT, {}, 0)
+end
```

Comment:
More of a style thing, but conditional timing shouldn't take place in functions with such self-explanatory names like this.
Also it's very important that we check current commands, so we don't erroneously unwait units that already have an active wait. This is an unlikely case given that waited units shouldn't be consuming energy, but I think it's better safe than sorry.
```lua
local function waitUnits()
	waitingUnits = getEnergyConsumingBuilders()
	local toWait = {}
	for i = 1, #waitingUnits do
		local uCmd = spGetUnitCurrentCommand(waitingUnits[i], 1)
		if uCmd ~= CMD_WAIT then
			toWait[#toWait + 1] = waitingUnits[i]
		end
	end
	Spring.GiveOrderToUnitArray(waitingUnits, CMD_WAIT, {}, 0)
end
```